core/src/main/scala/tapir/generic/ValidateMagnoliaDerivation.scala

```diff
@@ -0,0 +1,80 @@
+package tapir.generic
+
+import magnolia.{CaseClass, Magnolia, SealedTrait}
+
+trait ValidateMagnoliaDerivation extends LowPriorityValidators {
+  type Typeclass[T] = Validator[T]
+
+  def combine[T](ctx: CaseClass[Validator, T]): Validator[T] = {
+    ProductValidator(ctx.parameters.map { p =>
+      p.label -> new FieldValidator[T] {
+        override type fType = p.PType
+        override def get(t: T): fType = { p.dereference(t) }
+        override def validator: Typeclass[fType] = p.typeclass
+      }
+    }.toMap)
+  }
+
+  def dispatch[T](ctx: SealedTrait[Validator, T]): Validator[T] = Validator.rejecting
+
+  implicit def gen[T]: Validator[T] = macro Magnolia.gen[T]
+}
+
+trait LowPriorityValidators {
+  def fallback[T]: Validator[T] = Validator.passing
+}
+
+trait Validator[T] { outer =>
+  def validate(t: T): Boolean
+  def map[TT](g: TT => T): Validator[TT] = (t: TT) => {
+    outer.validate(g(t))
+  }
+  def forOption: Validator[Option[T]] = (t: Option[T]) => {
+    t.forall(outer.validate)
+  }
+}
+
+object Validator extends ValidateMagnoliaDerivation {
+  def passing[T]: Validator[T] = (t: T) => true
+  def rejecting[T]: Validator[T] = (t: T) => false
+}
+
+case class ProductValidator[T](fields: Map[String, FieldValidator[T]]) extends Validator[T] {
+  override def validate(t: T): Boolean = {
+    fields.values.forall { f =>
+      f.validator.validate(f.get(t))
+    }
+  }
+}
+
+trait FieldValidator[T] {
+  type fType
+  def get(t: T): fType
+  def validator: Validator[fType]
+}
+
+case class ValueValidator[T](constraints: List[Constraint[T]]) extends Validator[T] {
+  override def validate(t: T): Boolean = constraints.forall(_.check(t))
+}
+
+trait Constraint[T] {
```

Comment:
Will this eventually be `sealed`, or users be able to define their own custom constraints? If it's custom, I think there will need to be a 2nd method defining how to express the constraint in the OpenAPI model.

Reply:
I think we should allow users to define their custom constraints. Initially we might only support a well-defined subset for openapi (that's a closed set anyway), and discard any custom constraints

Reply:
For the first version lets make it `sealed` and we will think about allowing users to define their custom constraints latter.

Reply:
maybe let's start with a simple `ValidationError(msg: String)`?